src/metabase/models/action.clj

```diff
@@ -203,41 +211,43 @@
-        models-with-implicit-actions    (if known-models
-                                          (->> known-models
-                                               (filter #(contains? implicit-action-model-ids (:id %)))
-                                               distinct)
-                                          (when (seq implicit-action-model-ids)
-                                            (db/select 'Card :id [:in implicit-action-model-ids])))
-        implicit-parameters-by-model-id (when (seq models-with-implicit-actions)
-                                          (implicit-action-parameters models-with-implicit-actions))]
-    (for [{:keys [parameters] :as action} actions
-          :let [model-id        (:model_id action)
-                implicit-params (when (= (:type action) :implicit)
-                                  (let [implicit-params (get implicit-parameters-by-model-id model-id)
-                                        saved-params    (m/index-by :id parameters)]
-                                    (for [param implicit-params
-                                          :let [saved-param (get saved-params (:id param))]]
-                                      (merge param saved-param))))]]
-      (cond-> action
-        implicit-params
-        (m/assoc-some :parameters (cond->> implicit-params
-                                    (= "row/delete" (:kind action))
-                                    (filter ::pk?)
-
-                                    (contains? #{"row/update" "row/delete"} (:kind action))
-                                    (map (fn [param] (cond-> param (::pk? param) (assoc :required true))))
-
-                                    :always
-                                    (map #(dissoc % ::pk? ::field-id))
-
-                                    :always seq))))))
+  (let [actions                       (apply select-actions-without-implicit-params options)
+        implicit-action-model-ids     (set (map :model_id (filter (comp #(= :implicit %) :type) actions)))
+        implicit-action-models        (if known-models
+                                        (->> known-models
+                                             (filter #(contains? implicit-action-model-ids (:id %)))
+                                             distinct)
+                                        (when (seq implicit-action-model-ids)
+                                          (db/select 'Card :id [:in implicit-action-model-ids])))
+        model-id->db-id               (into {} (for [card implicit-action-models]
+                                                 [(:id card) (:database_id card)]))
+        model-id->implicit-parameters (when (seq implicit-action-models)
+                                        (implicit-action-parameters implicit-action-models))]
+    (for [{:keys [parameters] :as action} actions]
+      (if (= (:type action) :implicit)
```

Comment:
man, it's dumb, but around this point I'm getting a little antsy for a `(defn implicit? [{action-type :type}] (= :implicit action-type))` helper

Reply:
eh, i'm not that antsy